test/metabase/query_processor_test/join_datetime_cast_test.clj

```diff
@@ -0,0 +1,59 @@
+(ns metabase.query-processor-test.join-datetime-cast-test
+  "Tests for joins with columns cast to datetime (issue #62099)"
+  (:require
+   [clojure.test :refer :all]
+   [metabase.lib.core :as lib]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor :as qp]
+   [metabase.test :as mt]))
+
+(deftest join-datetime-cast-test
+  (testing "Joining columns that are cast to Datetime should work (#62099)"
+    (mt/test-drivers (mt/normal-drivers-with-feature :left-join :expressions/datetime)
+      (let [mp (mt/metadata-provider)]
+        (testing "Join two tables on datetime cast fields"
+          (let [query (mt/mbql-query orders
+                        {:joins [{:source-table $$checkins
+                                  :alias "c"
+                                  :condition [:=
+                                              [:datetime $created_at]
+                                              [:datetime &c.checkins.date]]
+                                  :fields :all}]
+                         :limit 5})]
+            (mt/with-native-query-testing-context query
+              (is (some? (qp/process-query query))
+                  "Query with datetime cast join should not fail"))))
```

Comment:
This does not assert that the query does not fail, because if it does it still returns something -- an error message.